enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,616 @@
+---
+title: microshift-updateability-ostree
+authors:
+  - "@pmtk"
+reviewers: # Include a comment about what domain expertise a reviewer is expected to bring and what area of the enhancement you expect them to focus on. For example: - "@networkguru, for networking aspects, please look at IP bootstrapping aspect"
+  - "@pacevedom, MicroShift team"
+  - "@ggiguash, MicroShift team"
+  - "@copejon, MicroShift team"
+  - "@dusk125, etcd team"
+  - "@oglok, author of previous enhancement"
+  - "RHEL expert"
+approvers: # A single approver is preferred, the role of the approver is to raise important questions, help ensure the enhancement receives reviews from all applicable areas/SMEs, and determine when consensus is achieved such that the EP can move forward to implementation.  Having multiple approvers makes it difficult to determine who is responsible for the actual approval.
+  - "@dhellmann"
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - None
+creation-date: 2023-04-14
+last-updated: 2023-04-14
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/USHIFT-518
+see-also:
+  - "/enhancements/microshift/microshift-greenboot.md"
+  - "/enhancements/microshift/etcd-supportability.md"
+replaces:
+  - https://github.com/openshift/enhancements/pull/1312
+superseded-by:
+  - None
+---
+
+# MicroShift updateability in ostree based systems
+
+## Summary
+
+This enhancement focuses on high level overview of updating
+MicroShift running on ostree based systems such as RHEL 4 Edge.
+Enhancement covers backup and restore of MicroShift data,
+version migration (upgrade and downgrade) of MicroShift and 
+its consequences (migration of data between schema versions),
+and interactions with GreenBoot and operating system.
+
+## Motivation
+
+MicroShift is working towards a general availability (GA) release.
+As GA product, it is expected that it can be updated to
+provide security patches, functional updates, and bug fixes
+without needing to redeploy.
+
+MicroShift is intended to be a part of Red Hat Device Edge
+which is based on RHEL For Edge which features ostree and as such
+provides rollbacks to go back to previous ostree deployment.
+Even though, OpenShift does not support downgrade or rollback,
+MicroShift must support it in some form. Explicit downgrade 
+will not be supported, and rollback will be supported only when
+newer ostree commits is unhealthy and backup consistent with previously
+ran MicroShift is present.
```

Comment:
What are we planning to do when an sysadm chooses to boot the system on the previous version? 
It is as simple as running a ostree rollback operation.
In this case, greenboot may report "green" status, but system roolback will still take place.

Reply:
It's mentioned below. From our perspective this is a downgrade which is unsupported.
Rollback is doable when backup for older version wasn't overwritten

Reply:
Well, this may be a bit problematic because we're "losing" the previous backup on a successful boot. So, the operation would only succeed "sometimes".

To mitigate, what do you think to tie a backup with a ostree deployment, .i.e. `/var/lib/microshift.back.<deploymentID>`

We can query `rpm-ostree` for this information:
```
$ rpm-ostree status --json | jq '.deployments[] | .id + " " + .checksum + " " + if has("base-checksum") then .["base-checksum"] else "none" end'
"rhel-8497faf62210000ffb5274c8fb159512fd6b9074857ad46820daa1980842d889.0 8497faf62210000ffb5274c8fb159512fd6b9074857ad46820daa1980842d889 ffd58f65d6e7dee7851daad5908e5739aa1830aca20785d459f9e762885bc56d"
"rhel-ffd58f65d6e7dee7851daad5908e5739aa1830aca20785d459f9e762885bc56d.0 ffd58f65d6e7dee7851daad5908e5739aa1830aca20785d459f9e762885bc56d none"
```

Reply:
This goes back to decision to only keep only one backup to keep space reqs minimal.
And yes, that means that healthy boot of commit 2 would overwrite commit 1 backup, so if commit 2 begins to fails on subsequent boots, then even if we rollback, we don't have anything to restore from

Reply:
We could implement a cleanup procedure in "green" scripts. It boils down to deleting backups for non-existent deployments. Anyway, this is my recommendation based on my previous experience from ostree/greenboot work.
If we're still adamant to have a single backup, let's make it work the best we can with the limitations.